source/particle/generator/reference_cell.cc

```diff
@@ -0,0 +1,192 @@
+/*
+  Copyright (C) 2015 by the authors of the ASPECT code.
+
+ This file is part of ASPECT.
+
+ ASPECT is free software; you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation; either version 2, or (at your option)
+ any later version.
+
+ ASPECT is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with ASPECT; see the file doc/COPYING.  If not see
+ <http://www.gnu.org/licenses/>.
+ */
+
+#include <aspect/particle/generator/reference_cell.h>
+
+namespace aspect
+{
+  namespace Particle
+  {
+    namespace Generator
+    {
+      template <int dim>
+      ReferenceCell<dim>::ReferenceCell()
+      {
+        starting_particle_index = 0;
+      }
+
+      template <int dim>
+      void
+      ReferenceCell<dim>::generate_particles(std::multimap<types::LevelInd, Particle<dim> > &particles)
+      {
+        const std::vector<Point<dim>> particles_in_unit_cell = generate_particle_positions_in_unit_cell();
+
+        unsigned int n_particles_to_generate = this->get_triangulation().n_locally_owned_active_cells() * particles_in_unit_cell.size();
+        unsigned int prefix_sum = 0;
+
+        MPI_Scan(&n_particles_to_generate, &prefix_sum, 1, MPI_UNSIGNED, MPI_SUM, this->get_mpi_communicator());
+
+        starting_particle_index += prefix_sum - n_particles_to_generate;
+        types::particle_index particle_index = starting_particle_index;
+        typename Triangulation<dim>::active_cell_iterator
+        cell = this->get_triangulation().begin_active(),
+        endc = this->get_triangulation().end();
+
+        for (; cell != endc; cell++)
+          {
+            if (cell->is_locally_owned())
+              {
+                for (typename std::vector<Point<dim>>::const_iterator itr_particles_in_unit_cell = particles_in_unit_cell.begin();
+                     itr_particles_in_unit_cell != particles_in_unit_cell.end();
+                     itr_particles_in_unit_cell++)
+                  {
+                    try
```

Comment:
You can remove the try-catch block. As long as your `generate_particle_positions_in_unit_cell` function does its job correctly, there should not be any problem (and if it is not, there should be an assertion inside that function that the generated points lie inside the unit cell). The interior of the try block is correct.